File: infrastructures/ddd/typescript/src/api/middleware/CorrelationIdMiddleware.ts
import { Request, Response, NextFunction } from 'express';
import { randomUUID } from 'crypto';

/**
 * Middleware for tracking requests across services using correlation IDs.
 * MUST be applied FIRST in the middleware pipeline.
 * 
 * This middleware:
 * - Extracts correlation ID from incoming request headers
 * - Generates a new one if not present
 * - Adds it to request headers for downstream use
 * - Adds it to response headers for client tracking
 * 
 * Header name: x-correlation-id
 * 
 * Middleware order:
 * 1. Correlation ID (THIS) → 2. Logging → 3. Authentication → 
 * 4. Authorization/Ownership → 5. UnitOfWork → 6. Controller →
 * 7. UnitOfWork commit/rollback → 8. Error Handling
 */
export class CorrelationIdMiddleware {
  /**
   * Middleware handler function.
   * Use with Express: app.use((req, res, next) => CorrelationIdMiddleware.Handle(req, res, next))
   */
  public static Handle(p_request: Request, p_response: Response, p_next: NextFunction): void {
    // Extract correlation ID from header or generate new one
    let correlationId = p_request.headers['x-correlation-id'] as string;
    
    if (!correlationId) {
      correlationId = randomUUID();
    }

    // Add correlation ID to request headers for downstream use
    p_request.headers['x-correlation-id'] = correlationId;

    // Add correlation ID to response headers for client tracking
    p_response.setHeader('x-correlation-id', correlationId);

    p_next();
  }
}

File: infrastructures/ddd/csharp/src/Api/Middleware/CorrelationIdMiddleware.cs
using System;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;

namespace OddlyDdd.Api.Middleware
{
    /// <summary>
    /// Middleware for managing correlation IDs across requests.
    /// MUST be applied FIRST in the middleware pipeline.
    /// Extracts or generates a correlation ID for tracking requests across services.
    /// </summary>
    public class CorrelationIdMiddleware
    {
        private readonly RequestDelegate _next;
        private const string CorrelationIdHeader = "X-Correlation-Id";

        public CorrelationIdMiddleware(RequestDelegate p_next)
        {
            _next = p_next;
        }

        public async Task InvokeAsync(HttpContext p_context)
        {
            string correlationId;

            // Try to get correlation ID from header
            if (p_context.Request.Headers.TryGetValue(CorrelationIdHeader, out var headerValue))
            {
                correlationId = headerValue.ToString();
            }
            else
            {
                // Generate new correlation ID if not provided
                correlationId = Guid.NewGuid().ToString();
            }

            // Store in context for later retrieval
            p_context.Items["CorrelationId"] = correlationId;

            // Add to response headers
            p_context.Response.Headers[CorrelationIdHeader] = correlationId;

            await _next(p_context);
        }
    }
}
